File: apps/mobile/modules/location-engine/ios/LocationEngine.swift
import CoreLocation
import CoreMotion
import Foundation
import UIKit

/**
 * The iOS background-location engine (spec §9).
 *
 * Strategy, and why:
 *
 *  - **Significant-change monitoring is the default.** It is the only mechanism
 *    that reliably relaunches a terminated app, and it costs almost no battery
 *    because it rides cell/Wi-Fi transitions the radio is already observing.
 *  - **Region monitoring** provides arrival and departure at saved places.
 *  - **Visit monitoring** gives high-quality "arrived somewhere and stayed"
 *    events that significant-change alone misses.
 *  - **Standard location updates run only during a live session**, and stop
 *    automatically when it expires. Navigation-grade GPS is never left running:
 *    it would flatten a phone in hours and is not needed to answer "is my
 *    family member home yet".
 *
 * This engine makes no promise about update frequency, and the API deliberately
 * exposes none. iOS decides when to deliver significant-change and visit
 * events; the honest contract is "best effort, target freshness", which is what
 * the product copy says too.
 */
final class LocationEngine: NSObject {
    static let shared = LocationEngine()

    enum LiveSessionEndReason {
        case expired
        case stoppedByUser
        case criticalBattery
        case permissionLost
    }

    /// Set by the Expo module so the engine can push events to JavaScript.
    var eventSink: ((String, [String: Any?]) -> Void)?

    private let manager = CLLocationManager()
    private let motionActivityManager = CMMotionActivityManager()
    private let queue = LocationEventQueue()

    private var configuration = EngineConfiguration()
    private var state: TrackingState = .disabled
    private var sharingPaused = false
    private var geofences: [GeofenceRegion] = []
    private var lastAcceptedLocation: CLLocation?
    private var lastUploadAttemptAt: Date?
    private var lastUploadError: String?

    private var liveSessionId: String?
    private var liveSessionTimer: Timer?

    override private init() {
        super.init()
        manager.delegate = self
        manager.pausesLocationUpdatesAutomatically = true
        // Non-negotiable: while a live session is running the OS must show the
        // blue indicator. Hiding it would make this covert tracking (spec §9).
        manager.showsBackgroundLocationIndicator = true
        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    // MARK: - Configuration

    func configure(_ configuration: EngineConfiguration) throws {
        self.configuration = configuration
        applyPosture()
        emitState(reason: "configured")
    }

    // MARK: - Lifecycle

    func startPassiveTracking() throws {
        guard !sharingPaused else {
            // Resuming is an explicit, user-initiated act. Silently starting here
            // would let a caller override a pause the user asked for.
            throw EngineError.sharingPaused
        }
        guard CLLocationManager.locationServicesEnabled() else {
            transition(to: .permissionRequired, reason: "location-services-disabled")
            throw EngineError.locationServicesDisabled
        }

        switch manager.authorizationStatus {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
            transition(to: .permissionRequired, reason: "authorization-not-determined")
            throw EngineError.authorizationNotDetermined
        case .denied, .restricted:
            transition(to: .permissionRequired, reason: "authorization-denied")
            throw EngineError.authorizationDenied
        case .authorizedWhenInUse, .authorizedAlways:
            break
        @unknown default:
            transition(to: .permissionRequired, reason: "authorization-unknown")
            throw EngineError.authorizationDenied
        }

        if manager.authorizationStatus == .authorizedAlways {
            manager.allowsBackgroundLocationUpdates = true
            manager.startMonitoringSignificantLocationChanges()
            manager.startMonitoringVisits()
        }

        transition(to: batteryAdjustedAmbientState(), reason: "passive-started")
    }

    func pauseSharing() {
        sharingPaused = true
        stopLiveSession(reason: .stoppedByUser)
        manager.stopUpdatingLocation()
        manager.stopMonitoringSignificantLocationChanges()
        manager.stopMonitoringVisits()
        unregisterAllRegions()
        // Anything already captured but not yet uploaded is discarded: a user who
        // pauses expects nothing further to reach their family.
        queue.removeAll()
        transition(to: .disabled, reason: "sharing-paused")
    }

    func resumeSharing() throws {
        sharingPaused = false
        try startPassiveTracking()
        registerGeofencesInternal()
        emitState(reason: "sharing-resumed")
    }

    // MARK: - Live sessions

    func startLiveSession(sessionId: String, durationSeconds: Double, updateIntervalSeconds _: Double?) throws {
        guard !sharingPaused else { throw EngineError.sharingPaused }
        guard manager.authorizationStatus == .authorizedAlways
            || manager.authorizationStatus == .authorizedWhenInUse
        else {
            throw EngineError.authorizationDenied
        }
        // A live session is the most expensive thing the engine does; refusing it
        // on a nearly-dead battery keeps the device alive to send its last known
        // position, which matters more.
        if batteryLevel().map({ $0 <= configuration.criticalBatteryThreshold }) ?? false {
            throw EngineError.criticalBattery
        }

        stopLiveSession(reason: .stoppedByUser)

        let duration = min(
            EngineConfiguration.liveDurationRange.clamp(durationSeconds),
            configuration.liveSessionMaxSeconds
        )
        liveSessionId = sessionId
        transition(to: .live, reason: "live-session-started")

        manager.desiredAccuracy = TrackingState.live.desiredAccuracy
        manager.distanceFilter = configuration.distanceFilter(for: .live)
        manager.startUpdatingLocation()

        // Hard local deadline. The server also expires the session, but the device
        // must stop on its own even if it is offline when the session ends.
        liveSessionTimer?.invalidate()
        liveSessionTimer = Timer.scheduledTimer(withTimeInterval: duration, repeats: false) { [weak self] _ in
            self?.stopLiveSession(reason: .expired)
        }
    }

    func stopLiveSession(reason: LiveSessionEndReason) {
        guard liveSessionId != nil else { return }
        liveSessionTimer?.invalidate()
        liveSessionTimer = nil
        liveSessionId = nil
        manager.stopUpdatingLocation()
        transition(to: batteryAdjustedAmbientState(), reason: "live-session-ended:\(reason)")
    }

    // MARK: - Geofences

    func registerGeofences(_ regions: [GeofenceRegion]) throws {
        geofences = regions
        registerGeofencesInternal()
    }

    func unregisterGeofences(placeIds: [String]) {
        let removing = Set(placeIds)
        geofences.removeAll { removing.contains($0.placeId) }
        for region in manager.monitoredRegions where removing.contains(region.identifier) {
            manager.stopMonitoring(for: region)
        }
    }

    private func registerGeofencesInternal() {
        guard !sharingPaused,
              CLLocationManager.isMonitoringAvailable(for: CLCircularRegion.self) else { return }

        unregisterAllRegions()
        let selected = GeofenceRegion.prioritise(geofences, around: lastAcceptedLocation)
        let maximumRadius = manager.maximumRegionMonitoringDistance
        for region in selected {
            manager.startMonitoring(for: region.asCircularRegion(maximumRadius: maximumRadius))
        }
    }

    private func unregisterAllRegions() {
        for region in manager.monitoredRegions {
            manager.stopMonitoring(for: region)
        }
    }

    // MARK: - Reporting

    func permissionState() -> [String: Any?] {
        let status = manager.authorizationStatus
        let authorization = switch status {
        case .notDetermined: "NOT_DETERMINED"
        case .restricted: "RESTRICTED"
        case .denied: "DENIED"
        case .authorizedWhenInUse: "WHEN_IN_USE"
        case .authorizedAlways: "ALWAYS"
        @unknown default: "NOT_DETERMINED"
        }

        return [
            "authorization": authorization,
            "preciseLocationEnabled": manager.accuracyAuthorization == .fullAccuracy,
            "locationServicesEnabled": CLLocationManager.locationServicesEnabled(),
            "notificationsEnabled": NotificationAuthorization.isAuthorized,
            // On iOS this is Background App Refresh: without it, background delivery
            // is unreliable no matter what the location permission says.
            "backgroundRefreshEnabled": UIApplication.shared.backgroundRefreshStatus == .available,
            "foregroundServicePermissionGranted": nil,
            "batteryOptimizationIgnored": nil,
        ]
    }

    func deviceHealth() -> [String: Any?] {
        let formatter = ISO8601DateFormatter()
        return [
            "permission": permissionState(),
            "trackingState": state.rawValue,
            "batteryLevel": batteryLevel(),
            "isLowPowerMode": ProcessInfo.processInfo.isLowPowerModeEnabled,
            "isCharging": UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full,
            "pendingEventCount": queue.count,
            "oldestPendingEventAt": queue.oldestCapturedAt.map(formatter.string(from:)),
            "lastAcceptedAt": lastAcceptedLocation.map { formatter.string(from: $0.timestamp) },
            "lastUploadAttemptAt": lastUploadAttemptAt.map(formatter.string(from:)),
            "lastUploadError": lastUploadError,
            "remoteConfigVersion": configuration.configVersion,
        ]
    }

    /**
     * Hands the pending batch to JavaScript, which owns the HTTP call. The native
     * side keeps the queue; the TypeScript side keeps the retry policy, so both
     * halves are testable independently.
     */
    func flushPendingEvents() -> [String: Any?] {
        lastUploadAttemptAt = Date()
        let formatter = ISO8601DateFormatter()
        let batch = queue.peek(limit: configuration.uploadBatchSize)

        return [
            "uploadedCount": 0,
            "remainingCount": queue.count,
            "lastError": lastUploadError,
            "attemptedAt": formatter.string(from: Date()),
            "pending": batch.map { record in
                [
                    "eventId": record.eventId,
                    "sequenceNumber": record.sequenceNumber,
                    "latitude": record.latitude,
                    "longitude": record.longitude,
                    "horizontalAccuracy": record.horizontalAccuracy,
                    "altitude": record.altitude,
                    "speed": record.speed,
                    "heading": record.heading,
                    "batteryLevel": record.batteryLevel,
                    "isLowPowerMode": record.isLowPowerMode,
                    "trackingMode": record.trackingMode,
                    "capturedAt": formatter.string(from: record.capturedAt),
                    "createdAt": formatter.string(from: record.createdAt),
                ]
            },
        ]
    }

    func confirmUploaded(eventIds: [String]) {
        queue.remove(eventIds: Set(eventIds))
        lastUploadError = nil
        emitQueueChange()
    }

    // MARK: - Internals

    private func batteryLevel() -> Double? {
        let level = UIDevice.current.batteryLevel
        return level < 0 ? nil : Double(level)
    }

    /// The ambient state to fall back to, after accounting for battery.
    private func batteryAdjustedAmbientState() -> TrackingState {
        guard !sharingPaused else { return .disabled }
        if let level = batteryLevel() {
            if level <= configuration.criticalBatteryThreshold { return .criticalBattery }
            if level <= configuration.lowBatteryThreshold { return .lowBattery }
        }
        if ProcessInfo.processInfo.isLowPowerModeEnabled { return .lowBattery }
        return .passive
    }

    private func applyPosture() {
        manager.desiredAccuracy = state.desiredAccuracy
        manager.distanceFilter = configuration.distanceFilter(for: state)
        if !state.allowsContinuousUpdates {
            manager.stopUpdatingLocation()
        }
    }

    private func transition(to next: TrackingState, reason: String) {
        guard next != state else { return }
        state = next
        applyPosture()
        emitState(reason: reason)
    }

    private func emitState(reason: String) {
        eventSink?("onEngineStateChange", ["state": state.rawValue, "reason": reason])
    }

    private func emitQueueChange() {
        eventSink?("onQueueChange", ["pendingEventCount": queue.count])
    }

    /**
     * Acceptance gate. A fix that fails any of these is dropped on the device
     * rather than uploaded and rejected by the server — that saves radio, battery
     * and backend cost, and keeps obviously-wrong data out of the family map.
     *
     * No branch here logs a coordinate.
     */
    private func accept(_ location: CLLocation) {
        guard state.producesLocation, !sharingPaused else { return }

        // Negative accuracy is Core Location's "this fix is invalid".
        guard location.horizontalAccuracy >= 0,
              location.horizontalAccuracy <= configuration.maxAcceptableAccuracyMeters else { return }

        let age = -location.timestamp.timeIntervalSinceNow
        guard age <= LocationEventQueue.maxAgeSeconds, age >= -120 else { return }

        if let previous = lastAcceptedLocation {
            let distance = location.distance(from: previous)
            let elapsed = location.timestamp.timeIntervalSince(previous.timestamp)
            // Duplicate: same place, same moment.
            if distance < 20, elapsed < 60 { return }
            // Implausible ground speed implies a spoofed or corrupt fix.
            if elapsed > 0, distance / elapsed > 350 { return }
            // In stationary posture, do not persist repeated identical points.
            if state == .stationary, distance < configuration.distanceFilter(for: .stationary) { return }
        }

        lastAcceptedLocation = location

        queue.enqueue(
            LocationEventQueue.Record(
                eventId: UUID().uuidString,
                sequenceNumber: queue.nextSequenceNumber(),
                latitude: location.coordinate.latitude,
                longitude: location.coordinate.longitude,
                horizontalAccuracy: location.horizontalAccuracy,
                altitude: location.verticalAccuracy >= 0 ? location.altitude : nil,
                speed: location.speed >= 0 ? location.speed : nil,
                heading: location.course >= 0 ? location.course : nil,
                batteryLevel: batteryLevel(),
                isLowPowerMode: ProcessInfo.processInfo.isLowPowerModeEnabled,
                trackingMode: state.rawValue,
                capturedAt: location.timestamp,
                createdAt: Date()
            )
        )
        emitQueueChange()
    }

    enum EngineError: Error {
        case sharingPaused
        case locationServicesDisabled
        case authorizationDenied
        case authorizationNotDetermined
        case criticalBattery
    }
}

// MARK: - CLLocationManagerDelegate

extension LocationEngine: CLLocationManagerDelegate {
    func locationManager(_: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        locations.forEach(accept)
    }

    func locationManager(_: CLLocationManager, didVisit visit: CLVisit) {
        // A visit with a distant-past arrival is a departure record; either way the
        // coordinate is worth storing as a low-frequency anchor point.
        guard visit.horizontalAccuracy >= 0 else { return }
        let timestamp = visit.departureDate == Date.distantFuture ? visit.arrivalDate : visit.departureDate
        accept(
            CLLocation(
                coordinate: visit.coordinate,
                altitude: 0,
                horizontalAccuracy: visit.horizontalAccuracy,
                verticalAccuracy: -1,
                timestamp: timestamp
            )
        )
    }

    func locationManager(_: CLLocationManager, didEnterRegion region: CLRegion) {
        emitGeofence(region: region, transition: "ARRIVAL")
    }

    func locationManager(_: CLLocationManager, didExitRegion region: CLRegion) {
        emitGeofence(region: region, transition: "DEPARTURE")
    }

    private func emitGeofence(region: CLRegion, transition: String) {
        guard !sharingPaused else { return }
        eventSink?(
            "onGeofenceTransition",
            [
                // Opaque identifier only — never the coordinate of the place.
                "placeId": region.identifier,
                "transition": transition,
                "occurredAt": ISO8601DateFormatter().string(from: Date()),
            ]
        )
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        eventSink?("onPermissionChange", permissionState())

        switch manager.authorizationStatus {
        case .denied, .restricted, .notDetermined:
            // A downgrade at any time stops collection immediately.
            manager.stopUpdatingLocation()
            manager.stopMonitoringSignificantLocationChanges()
            manager.stopMonitoringVisits()
            unregisterAllRegions()
            stopLiveSession(reason: .permissionLost)
            transition(to: .permissionRequired, reason: "authorization-downgraded")
        case .authorizedAlways:
            manager.allowsBackgroundLocationUpdates = true
            manager.startMonitoringSignificantLocationChanges()
            manager.startMonitoringVisits()
            registerGeofencesInternal()
            transition(to: batteryAdjustedAmbientState(), reason: "authorization-always")
        case .authorizedWhenInUse:
            // Background delivery is not available; the app still works in the
            // foreground and the UI explains the limitation rather than hiding it.
            manager.allowsBackgroundLocationUpdates = false
            transition(to: batteryAdjustedAmbientState(), reason: "authorization-when-in-use")
        @unknown default:
            break
        }
    }

    func locationManager(_: CLLocationManager, didFailWithError error: Error) {
        // Sanitised: the reason is a code, never a coordinate or a full error dump.
        if let clError = error as? CLError {
            lastUploadError = "core-location-\(clError.code.rawValue)"
        } else {
            lastUploadError = "core-location-unknown"
        }
    }
}

/// Small indirection so the engine does not import UserNotifications eagerly.
private enum NotificationAuthorization {
    /// Updated by the app when it queries notification settings; defaults to
    /// false so the troubleshooting screen errs toward "check your settings".
    static var isAuthorized = false
}

File: apps/mobile/modules/location-engine/android/src/main/java/com/familylocation/locationengine/LocationEngine.kt
package com.familylocation.locationengine

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.location.Location
import android.os.Build
import com.google.android.gms.location.Geofence
import com.google.android.gms.location.GeofencingClient
import com.google.android.gms.location.GeofencingRequest
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import java.util.UUID

/**
 * The Android background-location engine (spec §9).
 *
 * Strategy, and why:
 *
 *  - **Fused Location Provider with a state-dependent priority.** Balanced
 *    power is the default; high accuracy is reserved for live sessions.
 *  - **Geofencing API** for arrival and departure at saved places. It is far
 *    cheaper than polling and keeps working while the app is not running.
 *  - **A foreground service only for live sessions**, with a visible
 *    notification the user cannot miss.
 *  - **WorkManager** for uploads, so a batch survives process death and waits
 *    for connectivity instead of burning the radio while offline.
 *
 * No promise is made about update frequency. Doze, App Standby and OEM battery
 * managers can all defer delivery, and the product copy says so rather than
 * claiming an interval the platform will not honour.
 */
class LocationEngine private constructor(context: Context) {

    private val appContext = context.applicationContext
    private val fusedClient = LocationServices.getFusedLocationProviderClient(appContext)
    private val geofencingClient: GeofencingClient = LocationServices.getGeofencingClient(appContext)
    private val permissions = PermissionMonitor(appContext)
    private val deviceHealth = DeviceHealth(appContext)
    private val queue = EncryptedEventQueue(appContext)

    private var configuration = EngineConfiguration()
    private var state: TrackingState = TrackingState.DISABLED
    private var sharingPaused = false
    private var lastAccepted: Location? = null
    private var activeLiveSessionId: String? = null
    private var registeredPlaceIds = mutableSetOf<String>()

    /** Set by the Expo module so the engine can push events to JavaScript. */
    var eventSink: ((String, Map<String, Any?>) -> Unit)? = null

    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
            result.locations.forEach(::accept)
        }
    }

    // MARK: - Configuration and lifecycle

    fun configure(config: EngineConfiguration) {
        configuration = config
        emitState("configured")
    }

    @SuppressLint("MissingPermission")
    fun startPassiveTracking() {
        check(!sharingPaused) { "Sharing is paused; resume it explicitly before tracking." }
        check(permissions.hasFineLocation() || permissions.hasCoarseLocation()) {
            "Location permission has not been granted."
        }

        transition(batteryAdjustedAmbientState(), "passive-started")
        requestUpdates(state)
        setSharingPersisted(true)
    }

    fun pauseSharing() {
        sharingPaused = true
        stopLiveSession()
        fusedClient.removeLocationUpdates(locationCallback)
        unregisterAllGeofences()
        // A user who pauses expects nothing further to reach their family, so
        // anything captured but not yet uploaded is discarded.
        queue.clear()
        setSharingPersisted(false)
        transition(TrackingState.DISABLED, "sharing-paused")
    }

    fun resumeSharing() {
        sharingPaused = false
        startPassiveTracking()
        emitState("sharing-resumed")
    }

    fun restoreAfterBoot() {
        if (sharingPaused || !permissions.canCollectInBackground()) return
        runCatching { startPassiveTracking() }
    }

    // MARK: - Live sessions

    fun startLiveSession(sessionId: String, durationSeconds: Long) {
        check(!sharingPaused) { "Sharing is paused." }
        val battery = deviceHealth.batteryLevel()
        // Refusing on a nearly-dead battery keeps the device alive long enough
        // to report its last known position, which matters more.
        check(battery == null || battery > configuration.criticalBatteryThreshold) {
            "Battery is too low to start a live session."
        }

        activeLiveSessionId = sessionId
        val capped = durationSeconds.coerceAtMost(configuration.liveSessionMaxSeconds.toLong())
        LiveSessionService.start(appContext, sessionId, capped)
        transition(TrackingState.LIVE, "live-session-started")
        requestUpdates(TrackingState.LIVE)
    }

    fun stopLiveSession() {
        if (activeLiveSessionId == null) return
        activeLiveSessionId = null
        LiveSessionService.stop(appContext)
        transition(batteryAdjustedAmbientState(), "live-session-stopped")
        requestUpdates(state)
    }

    fun onLiveSessionExpired(sessionId: String) {
        if (activeLiveSessionId != sessionId) return
        activeLiveSessionId = null
        transition(batteryAdjustedAmbientState(), "live-session-expired")
        requestUpdates(state)
    }

    // MARK: - Geofences

    @SuppressLint("MissingPermission")
    fun registerGeofences(places: List<Map<String, Any?>>) {
        if (sharingPaused || !permissions.canCollectInBackground()) return

        val geofences = places.mapNotNull { place ->
            val placeId = place["placeId"] as? String ?: return@mapNotNull null
            val latitude = (place["latitude"] as? Number)?.toDouble() ?: return@mapNotNull null
            val longitude = (place["longitude"] as? Number)?.toDouble() ?: return@mapNotNull null
            val radius = (place["radiusMeters"] as? Number)?.toFloat() ?: return@mapNotNull null

            var transitions = 0
            if (place["notifyOnArrival"] as? Boolean != false) {
                transitions = transitions or Geofence.GEOFENCE_TRANSITION_ENTER
            }
            if (place["notifyOnDeparture"] as? Boolean != false) {
                transitions = transitions or Geofence.GEOFENCE_TRANSITION_EXIT
            }
            if (transitions == 0) return@mapNotNull null

            registeredPlaceIds.add(placeId)
            Geofence.Builder()
                .setRequestId(placeId)
                .setCircularRegion(latitude, longitude, radius)
                .setExpirationDuration(Geofence.NEVER_EXPIRE)
                .setTransitionTypes(transitions)
                // Suppresses the burst of transitions a fresh registration would
                // otherwise emit for places the user is already inside.
                .setNotificationResponsiveness(NOTIFICATION_RESPONSIVENESS_MS)
                .build()
        }

        if (geofences.isEmpty()) return

        val request = GeofencingRequest.Builder()
            .setInitialTrigger(0)
            .addGeofences(geofences)
            .build()

        geofencingClient.addGeofences(request, geofencePendingIntent())
    }

    fun unregisterGeofences(placeIds: List<String>) {
        if (placeIds.isEmpty()) return
        registeredPlaceIds.removeAll(placeIds.toSet())
        geofencingClient.removeGeofences(placeIds)
    }

    private fun unregisterAllGeofences() {
        if (registeredPlaceIds.isEmpty()) return
        geofencingClient.removeGeofences(registeredPlaceIds.toList())
        registeredPlaceIds.clear()
    }

    private fun geofencePendingIntent(): PendingIntent {
        val intent = Intent(appContext, GeofenceReceiver::class.java)
        var flags = PendingIntent.FLAG_UPDATE_CURRENT
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            flags = flags or PendingIntent.FLAG_MUTABLE
        }
        return PendingIntent.getBroadcast(appContext, 0, intent, flags)
    }

    // MARK: - Reporting

    fun permissionState(): Map<String, Any?> = permissions.snapshot()

    fun health(): Map<String, Any?> = mapOf(
        "permission" to permissions.snapshot(),
        "trackingState" to state.wireName,
        "batteryLevel" to deviceHealth.batteryLevel(),
        "isLowPowerMode" to permissions.isPowerSaveMode(),
        "isCharging" to deviceHealth.isCharging(),
        "pendingEventCount" to queue.size,
        "oldestPendingEventAt" to queue.oldestCapturedAtMillis?.let(::iso8601),
        "lastAcceptedAt" to lastAccepted?.time?.let(::iso8601),
        "lastUploadAttemptAt" to null,
        "lastUploadError" to null,
        "remoteConfigVersion" to configuration.configVersion,
    )

    /** Hands the pending batch to JavaScript, which owns the HTTP call. */
    fun pendingBatch(): Map<String, Any?> {
        val batch = queue.peek(configuration.uploadBatchSize)
        return mapOf(
            "uploadedCount" to 0,
            "remainingCount" to queue.size,
            "lastError" to null,
            "attemptedAt" to iso8601(System.currentTimeMillis()),
            "pending" to batch.map { record ->
                mapOf(
                    "eventId" to record.eventId,
                    "sequenceNumber" to record.sequenceNumber,
                    "latitude" to record.latitude,
                    "longitude" to record.longitude,
                    "horizontalAccuracy" to record.horizontalAccuracy,
                    "altitude" to record.altitude,
                    "speed" to record.speed,
                    "heading" to record.heading,
                    "batteryLevel" to record.batteryLevel,
                    "isLowPowerMode" to record.isLowPowerMode,
                    "trackingMode" to record.trackingMode,
                    "capturedAt" to iso8601(record.capturedAtMillis),
                    "createdAt" to iso8601(record.createdAtMillis),
                )
            },
        )
    }

    fun confirmUploaded(eventIds: List<String>) {
        queue.remove(eventIds.toSet())
        emitQueueChange()
    }

    fun emitGeofenceTransition(placeId: String, transition: String, occurredAtMillis: Long) {
        if (sharingPaused) return
        eventSink?.invoke(
            "onGeofenceTransition",
            // Opaque identifier only — never the coordinate of the place.
            mapOf(
                "placeId" to placeId,
                "transition" to transition,
                "occurredAt" to iso8601(occurredAtMillis),
            ),
        )
    }

    fun reportGeofenceError(errorCode: Int) {
        eventSink?.invoke("onEngineStateChange", mapOf("state" to state.wireName, "reason" to "geofence-error-$errorCode"))
    }

    // MARK: - Internals

    @SuppressLint("MissingPermission")
    private fun requestUpdates(target: TrackingState) {
        fusedClient.removeLocationUpdates(locationCallback)
        if (!target.producesLocation) return

        val request = LocationRequest.Builder(target.priority, target.intervalMillis)
            .setMinUpdateDistanceMeters(configuration.distanceFilterFor(target))
            // Never wake the device more often than the ambient interval, even
            // if another app has a faster request running.
            .setMinUpdateIntervalMillis(target.intervalMillis)
            .setWaitForAccurateLocation(target == TrackingState.LIVE)
            .build()

        fusedClient.requestLocationUpdates(request, locationCallback, appContext.mainLooper)
    }

    private fun batteryAdjustedAmbientState(): TrackingState {
        if (sharingPaused) return TrackingState.DISABLED
        val level = deviceHealth.batteryLevel()
        if (level != null) {
            if (level <= configuration.criticalBatteryThreshold) return TrackingState.CRITICAL_BATTERY
            if (level <= configuration.lowBatteryThreshold) return TrackingState.LOW_BATTERY
        }
        if (permissions.isPowerSaveMode()) return TrackingState.LOW_BATTERY
        return TrackingState.PASSIVE
    }

    private fun transition(next: TrackingState, reason: String) {
        if (next == state) return
        state = next
        emitState(reason)
    }

    private fun emitState(reason: String) {
        eventSink?.invoke("onEngineStateChange", mapOf("state" to state.wireName, "reason" to reason))
    }

    private fun emitQueueChange() {
        eventSink?.invoke("onQueueChange", mapOf("pendingEventCount" to queue.size))
    }

    /**
     * Acceptance gate. A fix failing any of these is dropped on the device
     * rather than uploaded and rejected — that saves radio, battery and backend
     * cost, and keeps obviously-wrong points off the family map.
     *
     * No branch here logs a coordinate.
     */
    private fun accept(location: Location) {
        if (!state.producesLocation || sharingPaused) return
        if (!location.hasAccuracy() || location.accuracy < 0) return
        if (location.accuracy > configuration.maxAcceptableAccuracyMeters) return

        val ageMillis = System.currentTimeMillis() - location.time
        if (ageMillis > EncryptedEventQueue.MAX_AGE_MILLIS || ageMillis < -120_000) return

        lastAccepted?.let { previous ->
            val distance = location.distanceTo(previous)
            val elapsedSeconds = (location.time - previous.time) / 1000.0
            // Duplicate: same place, same moment.
            if (distance < 20f && elapsedSeconds < 60) return
            // Implausible ground speed implies a spoofed or corrupt fix.
            if (elapsedSeconds > 0 && distance / elapsedSeconds > 350) return
            if (state == TrackingState.STATIONARY &&
                distance < configuration.distanceFilterFor(TrackingState.STATIONARY)
            ) {
                return
            }
        }

        lastAccepted = location

        queue.enqueue(
            EncryptedEventQueue.Record(
                eventId = UUID.randomUUID().toString(),
                sequenceNumber = queue.nextSequenceNumber(),
                latitude = location.latitude,
                longitude = location.longitude,
                horizontalAccuracy = location.accuracy.toDouble(),
                altitude = if (location.hasAltitude()) location.altitude else null,
                speed = if (location.hasSpeed()) location.speed.toDouble() else null,
                heading = if (location.hasBearing()) location.bearing.toDouble() else null,
                batteryLevel = deviceHealth.batteryLevel(),
                isLowPowerMode = permissions.isPowerSaveMode(),
                trackingMode = state.wireName,
                capturedAtMillis = location.time,
                createdAtMillis = System.currentTimeMillis(),
            ),
        )
        emitQueueChange()
    }

    private fun setSharingPersisted(enabled: Boolean) {
        appContext.getSharedPreferences(BootReceiver.PREFS, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(BootReceiver.KEY_SHARING_ENABLED, enabled)
            .apply()
    }

    private fun iso8601(millis: Long): String =
        java.time.Instant.ofEpochMilli(millis).toString()

    companion object {
        private const val NOTIFICATION_RESPONSIVENESS_MS = 5 * 60 * 1000

        @Volatile
        var instance: LocationEngine? = null
            private set

        fun getOrCreate(context: Context): LocationEngine =
            instance ?: synchronized(this) {
                instance ?: LocationEngine(context).also { instance = it }
            }
    }
}
